scripts/abi_artifact_generator/templates/decoders.inc.jinja_template

```diff
@@ -20,103 +20,88 @@ This template file is distributed under the MIT License. See LICENSE.md for deta
-  MAKE_CONVERSION_HELPER(stack_state, state_ptr, stack_printing_helper);
-  PRINT_BYTES(stack_state, stack_printing_helper);
+  PRINT_AS_BYTES(stack_state, (stack_state *) data);
 }
 
-#define PRINT(TYPE, POINTER)                        \
-  do {                                              \
-    printf("      - Type: " #TYPE " # size = %zu\n" \
-           "        Bytes: ",                       \
-           sizeof(TYPE));                           \
-    MAKE_CONVERSION_HELPER(TYPE, POINTER, local_helper); \
-    PRINT_BYTES(TYPE, local_helper);                \
-  } while (0)
-
-#define POINTER(TYPE, POINTER)                        \
-  do {                                                \
-    printf("      - Type: " #TYPE " # size = %zu\n"   \
-           "        Pointer: 0x%zx\n        Bytes: ", \
-           sizeof(TYPE),                              \
-           (/*( register_type )*/) POINTER);          \
-    MAKE_CONVERSION_HELPER(TYPE *, POINTER, local_helper); \
-    PRINT_BYTES(TYPE *, local_helper);                \
+#define DECODE_A_SINGLE_VALUE(TYPE, SIZE, ADDRESS, INPUT, OUTPUT) \
+  do {                                                            \
+    printf("      - Type: " #TYPE " # size = %zu\n"               \
+           "        Address: 0x%0/*( register_size * 2 )*/zx\n"   \
+           "        AddressBytes: ",                              \
+           (/*( register_type )*/) (SIZE),                        \
+           (/*( register_type )*/) (ADDRESS));                    \
+    PRINT_AS_BYTES(/*( register_type )*/, &ADDRESS);              \
+    printf("        ExpectedBytes: ");                            \
+    print_bytes((INPUT), (SIZE));                                 \
+    printf("        FoundBytes: ");                               \
+    print_bytes((OUTPUT), (SIZE));                                \
   } while (0)
 
-#define ARGUMENT(TYPE, NAME, LOCATION, OFFSET) \
-  do {                                                                         \
-    TYPE NAME; (void) NAME;                                                    \
-    if (sizeof(TYPE) <= ARGUMENT_SIZE_LIMIT(TYPE, NAME) && !IS_PACKED(TYPE)) { \
-      PRINT(TYPE, (LOCATION) + (OFFSET));                                      \
-      (OFFSET) += sizeof(TYPE);                                                \
-    } else {                                                                   \
-      POINTER(TYPE, (LOCATION) + (OFFSET));                                    \
-      (OFFSET) += sizeof(TYPE *);                                              \
-    }                                                                          \
-  } while (0)
-
-ptrdiff_t current_offset = 0;
 /*{ for name, arguments in argument_functions.items() -}*/
-static void ABIDEF decode_/*( name )*/(const uint8_t *output) {
-  current_offset = 0;
+static void ABIDEF decode_/*( name )*/(const struct encoded value) {
```

Comment:
> Can you better describe what the `"Rework decoding to accept a full state"` commit does?

(cc: @aleclearmind)

I guess the name sucks indeed. The idea is to rework this interface to accept something a bit more structured as opposed to a stream of bytes. As in, to give a proper-ish interface to this header.